File: python/dmp_discrete/CanonicalSystemDiscrete.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 30 19:00:00 2017

@author: gsutanto
"""

import numpy as np
import os
import sys
import copy
sys.path.append(os.path.join(os.path.dirname(__file__), '../dmp_param/'))
sys.path.append(os.path.join(os.path.dirname(__file__), '../dmp_base/'))
sys.path.append(os.path.join(os.path.dirname(__file__), '../dmp_coupling/base/'))
from CanonicalSystem import *

class CanonicalSystemDiscrete(CanonicalSystem, object):
    'Class for canonical systems of discrete DMPs.'
    
    def __init__(self, tau_system, cs_order=2, cs_alpha=None, cs_beta=None, canonical_couplers_list=[], name=""):
        super(CanonicalSystemDiscrete, self).__init__(tau_system, canonical_couplers_list, name)
        self.order = cs_order
        assert ((self.order == 1) or (self.order == 2)), "Discrete canonical system order=" + str(self.order) + " is not supported! The only supported discrete canonical system order is either order==1 or order==2."
        if (cs_alpha != None):
            self.alpha = cs_alpha
        else:
            if (self.order == 2):
                self.alpha = 25.0
            elif (self.order == 1):
                self.alpha = 25.0/3.0
        if (cs_beta != None):
            self.beta = cs_beta
        else:
            if (self.order == 2):
                self.beta = self.alpha/4.0
            elif (self.order == 1):
                self.beta = 0.0
        self.x = 1.0
        self.xd = 0.0
        self.xdd = 0.0
        self.v = 0.0
        self.vd = 0.0
    
    def isValid(self):
        assert (super(CanonicalSystemDiscrete, self).isValid()), "CanonicalSystem is invalid!"
        assert ((self.order == 1) or (self.order == 2)), "Discrete canonical system order=" + str(self.order) + " is not supported! The only supported discrete canonical system order is either order==1 or order==2."
        assert (self.alpha > 0.0), "self.alpha=" + str(self.alpha) + " <= 0.0 (invalid!)"
        assert (self.beta >= 0.0), "self.beta=" + str(self.beta) + " < 0.0 (invalid!)"
        return True
    
    def start(self):
        assert (self.isValid()), "Pre-condition(s) checking is failed: this CanonicalSystemDiscrete is invalid!"
        self.x = 1.0
        self.xd = 0.0
        self.xdd = 0.0
        self.v = 0.0
        self.vd = 0.0
        self.resetCouplingTerm()
        self.is_started = True
        return None
    
    def getCanonicalPosition(self):
        return copy.copy(self.x)
    
    def getCanonicalVelocity(self):
        return copy.copy(self.v)
    
    def getCanonicalMultiplier(self):
        assert ((self.order == 1) or (self.order == 2)), "Discrete canonical system order=" + str(self.order) + " is not supported! The only supported discrete canonical system order is either order==1 or order==2."
        if (self.order == 1):
            return copy.copy(self.x)
        elif (self.order == 2):
            return copy.copy(self.v)
        else: # should NEVER reach here
            return None
    
    def updateCanonicalState(self, dt):
        assert (self.is_started), "CanonicalSystemDiscrete is NOT yet started!"
        assert (self.isValid()), "Pre-condition(s) checking is failed: this CanonicalSystemDiscrete is invalid!"
        assert (dt > 0.0), "dt=" + str(dt) + " <= 0.0 (invalid!)"
        
        tau = self.tau_sys.getTauRelative()
        C_c = self.getCouplingTerm()
        if (self.order == 2):
            self.xdd = self.vd * 1.0 / tau
            self.vd = ((self.alpha * ((self.beta * (0 - self.x)) - self.v)) + C_c) * 1.0 / tau
            self.xd = self.v * 1.0 / tau
        elif (self.order == 1):
            self.xdd = self.vd * 1.0 / tau
            self.vd = 0.0
            self.xd = ((self.alpha * (0 - self.x)) + C_c) * 1.0 / tau
        self.x = self.x + (self.xd * dt)
        self.v = self.v + (self.vd * dt)
        
        assert (self.x >= 0.0), "self.x=" + str(self.x) + " < 0.0 (invalid!)"
        return None

File: src/dmp_discrete/CanonicalSystemDiscrete.cpp
#include "dmp/dmp_discrete/CanonicalSystemDiscrete.h"

namespace dmp {

CanonicalSystemDiscrete::CanonicalSystemDiscrete()
    : CanonicalSystem(),
      order(2),
      alpha(25.0),
      beta(25.0 / 4.0),
      x(0),
      xd(0),
      xdd(0),
      v(0),
      vd(0) {}

CanonicalSystemDiscrete::CanonicalSystemDiscrete(
    TauSystem* tau_system, RealTimeAssertor* real_time_assertor, uint cs_order,
    std::vector<CanonicalCoupling*>* canonical_couplers)
    : CanonicalSystem(tau_system, real_time_assertor, canonical_couplers) {
  order = cs_order;
  if (order == 2) {
    alpha = 25.0;
    beta = alpha / 4.0;
  } else {
    order = 1;
    alpha = 25.0 / 3.0;
    beta = 0.0;
  }
  x = 1.0;
  xd = 0.0;
  xdd = 0.0;
  v = 0.0;
  vd = 0.0;
}

CanonicalSystemDiscrete::CanonicalSystemDiscrete(
    TauSystem* tau_system, uint cs_order, double cs_alpha,
    RealTimeAssertor* real_time_assertor,
    std::vector<CanonicalCoupling*>* canonical_couplers)
    : CanonicalSystem(tau_system, real_time_assertor, canonical_couplers) {
  order = cs_order;
  alpha = cs_alpha;
  if (order == 2) {
    beta = alpha / 4.0;
  } else {
    order = 1;
    beta = 0.0;
  }
  x = 1.0;
  xd = 0.0;
  xdd = 0.0;
  v = 0.0;
  vd = 0.0;
}

CanonicalSystemDiscrete::CanonicalSystemDiscrete(
    TauSystem* tau_system, uint cs_order, double cs_alpha, double cs_beta,
    RealTimeAssertor* real_time_assertor,
    std::vector<CanonicalCoupling*>* canonical_couplers)
    : CanonicalSystem(tau_system, real_time_assertor, canonical_couplers) {
  order = cs_order;
  alpha = cs_alpha;
  beta = cs_beta;
  x = 1.0;
  xd = 0.0;
  xdd = 0.0;
  v = 0.0;
  vd = 0.0;
}

bool CanonicalSystemDiscrete::isValid() {
  if (rt_assert(CanonicalSystem::isValid()) == false) {
    return false;
  }
  if (rt_assert((rt_assert(order >= 1)) && (rt_assert(order <= 2))) == false) {
    return false;
  }
  if (rt_assert(alpha > 0.0) == false) {
    return false;
  }
  if (rt_assert(beta >= 0.0) == false) {
    return false;
  }
  return true;
}

bool CanonicalSystemDiscrete::start() {
  if (rt_assert(CanonicalSystemDiscrete::isValid()) == false) {
    return false;
  }

  x = 1.0;
  xd = 0.0;
  xdd = 0.0;
  v = 0.0;
  vd = 0.0;

  is_started = true;

  return true;
}

uint CanonicalSystemDiscrete::getOrder() { return order; }

double CanonicalSystemDiscrete::getAlpha() { return alpha; }

double CanonicalSystemDiscrete::getBeta() { return beta; }

double CanonicalSystemDiscrete::getX() { return x; }

double CanonicalSystemDiscrete::getXd() { return xd; }

double CanonicalSystemDiscrete::getXdd() { return xdd; }

double CanonicalSystemDiscrete::getCanonicalPosition() { return x; }

double CanonicalSystemDiscrete::getCanonicalMultiplier() {
  if (order == 1) {
    return (x);
  } else if (order == 2) {
    return (v);
  } else {
    return 0.0;
  }
}

Vector3 CanonicalSystemDiscrete::getCanonicalState() {
  Vector3 canonical_state;
  canonical_state << x, xd, xdd;
  return (canonical_state);
}

bool CanonicalSystemDiscrete::updateCanonicalState(double dt) {
  if (rt_assert(is_started == true) == false) {
    return false;
  }
  if (rt_assert(CanonicalSystemDiscrete::isValid()) == false) {
    return false;
  }
  // input checking
  if (rt_assert(dt > 0.0) == false)  // dt does NOT make sense
  {
    return false;
  }

  double tau;
  if (rt_assert(tau_sys->getTauRelative(tau)) == false) {
    return false;
  }

  double C_c = 0.0;
  if (rt_assert(CanonicalSystem::getCouplingTerm(C_c)) == false) {
    return false;
  }

  if (order == 2) {
    xdd = vd / tau;
    vd = ((alpha * ((beta * (0 - x)) - v)) + C_c) / tau;
    xd = v / tau;
  } else {  // if (order == 1)
    xdd = vd / tau;
    vd = 0.0;  // vd  is irrelevant for 1st order canonical system, no need for
               // computation
    xd = ((alpha * (0 - x)) + C_c) / tau;
  }
  x = x + (xd * dt);
  v = v + (vd * dt);

  if (rt_assert(x >= 0.0) == false) {
    return false;
  }

  return true;
}

CanonicalSystemDiscrete::~CanonicalSystemDiscrete() {}

}  // namespace dmp
